syntax = "proto3";
package pulse.cdm.bind;
option java_package = "com.kitware.pulse.cdm.bind";
option csharp_namespace = "pulse.cdm.bind";
option optimize_for = SPEED;

/** @brief Data used to configure plots. */
message PlotData
{
  /** @brief Enumeration for percentage of baseline modes */
  enum ePercentageOfBaselineMode
  {
    Off                                                   = 0;  /**<< @brief No modification of values will occur. */
    All                                                   = 1;  /**<< @brief All values will be set to Value@TimeX / Values@Time0 */
    OnlyX                                                 = 2;  /**<< @brief Only X values will be set to Value@TimeX / Values@Time0 */
    OnlyY                                                 = 3;  /**<< @brief Only Y values will be set to Value@TimeX / Values@Time0 */
  }

  string                      DataFileOverride            = 1;  /**<< @brief Override the name of the results file to be loaded that was autogenned by the job name and the VerificationDir. */
  string                      DataPathValidationOverride  = 2;  /**<< @brief Set a relative path to the results file starting at the validation folder in the source tree. */
  optional double             FontSize                    = 3;  /**<< @brief Set the font size of axis labels and chart title (FontSize=22 by default). */
  bool                        Gridlines                   = 4;  /**<< @brief Turn gridlines on or off. */
  optional ImageDimensionData ImageDimensions             = 5;  /**<< @brief Specify the image width and height in pixels. Default is 1600x800, but ActionEventPlots can stretch automatically. */
  optional double             LegendFontSize              = 6;  /**<< @brief Set the font size of the legend (LegendFontSize=15 by default). */
  bool                        LogAxis                     = 7;  /**<< @brief Make all Y axes logarithmic. */
  string                      OutputFilename              = 8;  /**<< @brief Set the desired filename for the generated plot (will default to the chart title if nothing is set for this). */
  optional string             OutputOverride              = 9;  /**<< @brief Specify the path (relative to bin) to output the plot (../verification/Plots/ by default). */
  ePercentageOfBaselineMode   PercentOfBaselineMode       = 10;
  bool                        RemoveLegends               = 14; /**<< @brief Removes all legends from the generated plot. */
  optional string             Title                       = 15; /**<< @brief Will override the auto-generated plot title with the supplied string. */
  string                      VerificationDirectory       = 16; /**<< @brief Set the directory inside Scenarios in which to search for data (i.e. Patient, Drugs, Miscellaneous). */
  optional string             X1Label                     = 17; /**<< @brief Override the lower X axis label. */
  BoundsData                  X1Bounds                    = 18; /**<< @brief Specify the desired X1 bounds */
  optional string             Y1Label                     = 19; /**<< @brief Override the left-side Y axis label. */
  BoundsData                  Y1Bounds                    = 20; /**<< @brief Specify the desired Y1 bounds. */
}

/** @brief Data used to configure action event plots. */
message ActionEventPlotterData
{
  PlotData                    Plot                        = 1;

  repeated string             Header                      = 2;  /**<< @brief Specify the header(s) that you want to plot on the Y axis. */
  bool                        HideActionEventLegend       = 3;  /**<< @brief Will hide the legend for actions and events, leaving the legend for data and the action and event lines on the plot. */
  bool                        LegendOnly                  = 4;  /**<< @brief Generate an image with ONLY the legend, no plot included. */
  bool                        NoActions                   = 5;  /**<< @brief Actions will not be included on the plot or legend. */
  bool                        NoEvents                    = 6;  /**<< @brief Events will not be included on the plot or legend. */
  repeated string             OmitEventsWith              = 7;  /**<< @brief Will attempt to omit from the plot events that contain any of the specified case-sensitive words, i.e. OmitEventsWith=Tachycardia */
  repeated string             OmitActionsWith             = 8;  /**<< @brief Will attempt to omit from the plot actions that contain any of the specified case-sensitive words, i.e. OmitActionsWith=Hemorrhage,Infusion */
  string                      ValidationData              = 9;  /**<< @brief Specify the relative path and filename for validation results to be plotted as scatter points. */
  bool                        ZeroAxis                    = 10; /**<< @brief Will ensure that zero is included on the Y axis rather than zooming away from it based on the plot's min/max Y values. */
}

/** @brief Data used to configure convex hull plots. */
message ConvexHullPlotterData
{
  PlotData                    Plot                        = 1;
}

/** @brief Data used to configure multiplotter plots. */
message MultiPlotterData
{
  PlotData                    Plot                        = 1;

  repeated string             Y1                          = 2;  /**<< @brief Specify the header(s) to be plotted on the primary (left) Y axis. */
  repeated string             Y2                          = 3;  /**<< @brief Optional; Specify the header(s) to be plotted on the secondary (right) Y axis. */
  string                      X1                          = 4;  /**<< @brief Specify the header aginst which Y1 data will be plotted. */
  string                      X2                          = 5;  /**<< @brief Specify the header against which Y2 data will be plotted; required if you specify any Y2 headers. */
  BoundsData                  X2Bounds                    = 6;  /**<< @brief Specify the desired X2 bounds */
  BoundsData                  Y2Bounds                    = 7;  /**<< @brief Specify the desired Y2 bounds */
  optional string             Y2Label                     = 8;  /**<< @brief Override the right-side Y axis label. */
  optional string             X2Label                     = 9;  /**<< @brief Override the upper X axis label. */
  string                      ValidationData              = 10; /**<< @brief Specify the relative path and filename for experimental results to be plotted as an additional line. Whatever you enter for X2 and Y2 will be the headers pulled from the experimental data file. */
}

/** @brief Data used to configure multi-source plots. */
message MultiSourcePlotterData
{
  PlotData                    Plot                        = 1;

  repeated string             DataFileOverride            = 2; /**<< @brief NOTE: DataFileOverride within PlotData will be ignored. Override the name of the results file to be loaded that was autogenned by the job name and the VerificationDir. */
}

/** @brief Data used to specify bounds. */
message BoundsData
{
  optional double           LowerBound                  = 1;  /**<< @brief  Specify the desired lower bound. */
  optional double           UpperBound                  = 2;  /**<< @brief  Specify the desired upper bound. */
}

/** @brief Data used to specify image dimensions. */
message ImageDimensionData
{
  double                    Height_pixels               = 3;  /**<< @brief  Specify the desired image height in pixels. */
  double                    Width_pixels                = 4;  /**<< @brief  Specify the desired image width in pixels. */
}
